### PR TITLE
adopt apiVersionEnum which will filter out unnessary apiversions

### PR DESCRIPTION
**Pull request by chunyu3**
Fix https://github.com/microsoft/typespec/issues/4110

TCGC will filter out the apiVersions for sdk emitters, and give out an api-version Enum:

- filter out the api-version later than the default api-version specified

- filter out preview api-version when the target api-version is stable version.

This PR will use the tcgc apiVersion Enum type to do the api-version filter.

### CODE REVIEW

**Comment by chunyu3** [comment]
autorest.csharp regen preview pr: https://github.com/Azure/autorest.csharp/pull/4981

File: packages/http-client-csharp/emitter/src/sdk-context-options.ts [hunk 1]
```diff
@@ -1,7 +1,9 @@
 import { CreateSdkContextOptions } from "@azure-tools/typespec-client-generator-core";
 
 export let defaultSDKContextOptions: CreateSdkContextOptions = {
-  versioning: undefined,
+  versioning: {
```
Comment (chunyu3) [inline packages/http-client-csharp/emitter/src/sdk-context-options.ts:4]: By default, TCGC will following Azure preview/stable apiversion filter, MGC is common emitter, it should not follow this. So set the Regex which will not match any apiVersion to filter.

File: packages/http-client-csharp/emitter/src/lib/client-model-builder.ts [hunk 2]
```diff
@@ -34,7 +35,12 @@ export function createModel(sdkContext: SdkContext<NetEmitterOptions>): CodeMode
 
   navigateModels(sdkContext, modelMap, enumMap);
 
-  const rootApiVersions = getRootApiVersions(sdkPackage.clients);
+  const sdkApiVersionEnums = sdkPackage.enums.filter((e) => e.usage === UsageFlags.ApiVersionEnum);
+
+  const rootApiVersions =
+    sdkApiVersionEnums.length > 0
+      ? sdkApiVersionEnums[0].values.map((v) => v.value as string).flat()
+      : getRootApiVersions(sdkPackage.clients);
```
Comment (archerzz) [inline packages/http-client-csharp/emitter/src/lib/client-model-builder.ts:43]: If TCGC can guarantee that there will always be api version enums, then we can delete `getRootApiVersions` here.

File: packages/http-client-csharp/emitter/src/sdk-context-options.ts (hunk 1 above)
Comment (archerzz) [reply packages/http-client-csharp/emitter/src/sdk-context-options.ts:4]: Then what do we do in `autorest.csharp`? Shall we restore preview/stable api version filter?

Comment (chunyu3) [reply packages/http-client-csharp/emitter/src/sdk-context-options.ts:4]: We will set the filter back in autorest.csharp emitter. See the autorest.csharp regen preview pr

File: packages/http-client-csharp/emitter/src/lib/client-model-builder.ts (hunk 2 above)
Comment (chunyu3) [reply packages/http-client-csharp/emitter/src/lib/client-model-builder.ts:43]: yes. I add this incase there is no such api version enum.

**Review by archerzz — APPROVED** [review]

**Pull request merged**